### PR TITLE
Reopen: IDE Search Func UI Fix

### PR DESCRIPTION
**Pull request by hakankaradis**
Screencast of all ide search functionalities
http://recordit.co/vR55ER3f15

### CODE REVIEW

**Comment by hakankaradis** [comment]
Fixes: #9168

File: client/app/styl/commons/appfn.styl [hunk 1]
```diff
@@ -58,6 +58,20 @@ noScrollbar()
 bg(prop, args)
   background-{prop}         args
 
+msk(prop, args)
```
Comment (sinan) [inline client/app/styl/commons/appfn.styl]: 👍

File: client/app/styl/commons/appfn.styl [hunk 2]
```diff
@@ -235,6 +249,32 @@ r-sprite(folder, name)
...
+
+  requireName = s('~' + folder + '/sprites/vector/' + name + '.svg')
+
+  msk                       image, url(requireName)
+  msk                       position, center center
+  msk                       repeat, no-repeat
+  msk                       size, _width _height
+
+
+s-sprite(folder, name)
```
Comment (sinan) [inline client/app/styl/commons/appfn.styl]: 👍

File: client/ide/lib/views/contentsearch/idecontentsearchview.coffee [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+    @searchText     = searchText
+    @rootPath       = Encoder.XSSEncode @whereInput.getValue()
+    @selections     = @choices.getValue()
+
+    isCaseSensitive = @selections.indexOf('case-sensitive') > -1
+    isWholeWord     = @selections.indexOf('whole-word') > -1
+    isRegExp        = @selections.indexOf('regex') > -1
+    exts            = editorSettings.getAllExts()
+    include         = "\\*{#{exts.join ','}}"
+    exclureDirs     = Object.keys editorSettings.ignoreDirectories
```
Comment (sinan) [inline client/ide/lib/views/contentsearch/idecontentsearchview.coffee]: `excludeDirs` typo

File: client/ide/lib/views/contentsearch/idecontentsearchview.coffee [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+        @showWarning 'Something went wrong, please try again.'
+        kd.warn err
+
+    terms = @localStorageController.getValue('PreviousSearchTerms') or []
+
+    terms.push searchText
+    @localStorageController.setValue 'PreviousSearchTerms', terms
+
+
+  escapeRegExp: (str) -> str.replace /([.*+?\^${}()|\[\]\/\\])/g, '\\$1'
```
Comment (sinan) [inline client/ide/lib/views/contentsearch/idecontentsearchview.coffee]: keep all the regular expressions in a constant on top of the file above class description.

e.g.

```coffee
REGEXES =
  escapeRegExp: /([.*+?\^${}()|\[\]\/\\])/g
  escapeShell: /([\\"'`$\s\(\)<>])/g
…
```

File: client/ide/lib/views/contentsearch/idecontentsearchview.coffee [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+  grepEscapeRegExp: (str) -> str.replace /[[\]{}()*+?.,\\^$|#\s"']/g, '\\$&'
+
+
+  formatOutput: (output, callback = kd.noop) ->
+
+    return @showWarning 'Something went wrong, please try again.', yes  if output.stderr
+
+    # Regexes
+    mainLineRegex           = /^:?([\s\S]+):(\d+):([\s\S]*)$/
+    contextLineRegex        = /^([\s\S]+)\-(\d+)\-([\s\S]*)$/
```
Comment (sinan) [inline client/ide/lib/views/contentsearch/idecontentsearchview.coffee]: in the constant above.

File: client/ide/lib/views/contentsearch/idecontentsearchview.coffee [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+  createElements: ->
+
+    @findInput = new KDHitEnterInputView
+      type         : 'text'
+      cssClass     : 'search-input-with-icon'
+      placeholder  : 'Search in all files...'
+      validate     :
+        rules      :
+          required : yes
+      keydown      : _.bind @handleKeyDown, this
```
Comment (sinan) [inline client/ide/lib/views/contentsearch/idecontentsearchview.coffee]: `@bound ...`

File: client/ide/lib/views/contentsearch/idecontentsearchview.coffee [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+
+    @whereInput = new KDHitEnterInputView
+      type         : 'text'
+      cssClass     : 'search-input'
+      placeholder  : @rootPath
+      defaultValue : @rootPath
+      validate     :
+        rules      :
+          required : yes
+      keydown      : _.bind @handleKeyDown, this
```
Comment (sinan) [inline client/ide/lib/views/contentsearch/idecontentsearchview.coffee]: `@bound …`

File: client/ide/lib/views/contentsearch/idecontentsearchview.coffee [hunk 8]
```diff
@@ -0,0 +1,292 @@
+kd                          = require 'kd'
+KDButtonView                = kd.ButtonView
+KDCustomHTMLView            = kd.CustomHTMLView
+KDHitEnterInputView         = kd.HitEnterInputView
+KDMultipleChoice            = kd.MultipleChoice
+JView                       = require 'app/jview'
+_                           = require 'lodash'
```
Comment (sinan) [inline client/ide/lib/views/contentsearch/idecontentsearchview.coffee]: not necessary

File: client/ide/lib/views/filefinder/idefilefinder.coffee [hunk 9]
```diff
@@ -28,12 +28,12 @@ class IDEFileFinder extends KDCustomHTMLView
 
     @addSubView @input = input = new KDInputView
       type         : 'text'
-      placeholder  : 'Type a file name to search'
+      placeholder  : 'Type a filename to search...'
       keydown      : _.bind @handleKeyDown, this
       keyup        : _.bind @handleKeyUp, this
```
Comment (sinan) [inline client/ide/lib/views/filefinder/idefilefinder.coffee]: `@bound` also remove lodash unless used other places.

**Review by sinan — CHANGES_REQUESTED** [review]
@hakankaradis can you address the comments please.

File: client/ide/lib/views/contentsearch/idecontentsearchview.coffee [hunk 10]
```diff
@@ -0,0 +1,298 @@
...
+    return @destroy()  if key is 'esc'
+
+    if e.target is @findInput.getElement()
+      if key in [ 'up', 'down' ]
+        @showPreviousSearchLocations key
+
+
+  destroy: ->
+
+    uber = JView::destroy.bind this
```
Comment (sinan) [inline client/ide/lib/views/contentsearch/idecontentsearchview.coffee]: remove

File: client/ide/lib/views/contentsearch/idecontentsearchview.coffee [hunk 11]
```diff
@@ -0,0 +1,298 @@
...
+IDEContentSearchResultView  = require './idecontentsearchresultview'
+showError                   = require 'app/util/showError'
+showNotification            = require 'app/util/showNotification'
+
+REGEXES =
+  escapeShell      : /([\\"'`$\s\(\)<>])/g
+  escapeRegExp     : /([.*+?\^${}()|\[\]\/\\])/g
+  mainLineRegex    : /^:?([\s\S]+):(\d+):([\s\S]*)$/
+  grepEscapeRegExp : /[[\]{}()*+?.,\\^$|#\s"']/g
+  contextLineRegex : /^([\s\S]+)\-(\d+)\-([\s\S]*)$/
```
Comment (sinan) [inline client/ide/lib/views/contentsearch/idecontentsearchview.coffee]: pls be consistent either call it `regex` or `regexp`, `regex` is more widely used.

File: client/ide/lib/views/filefinder/idefilefinder.coffee [hunk 12]
```diff
@@ -28,12 +28,12 @@ class IDEFileFinder extends KDCustomHTMLView
 
     @addSubView @input = input = new KDInputView
       type         : 'text'
-      placeholder  : 'Type a file name to search'
-      keydown      : _.bind @handleKeyDown, this
-      keyup        : _.bind @handleKeyUp, this
+      placeholder  : 'Type a filename to search...'
```
Comment (sinan) [inline client/ide/lib/views/filefinder/idefilefinder.coffee]: use ellipsis `…` instead of three dots `...` tip: `opt + ;`

**Review by sinan — CHANGES_REQUESTED** [review]

**Comment by hakankaradis** [comment]
@sinan pr is updated.

File: client/ace/lib/acefindandreplaceview.coffee [hunk 13]
```diff
@@ -129,27 +129,26 @@ module.exports = class AceFindAndReplaceView extends JView
 
     @findInput = new KDHitEnterInputView
       type         : 'text'
-      placeholder  : 'Find...'
+      cssClass     : 'search-input-with-icon'
+      placeholder  : 'Find in the file…'
       validate     :
         rules      :
           required : yes
       keydown      : _.bind @handleKeyDown, this, yes
```
Comment (sinan) [inline client/ace/lib/acefindandreplaceview.coffee]: lodash^

**Review by sinan — APPROVED** [review]
this is from earlier commits I see, you can fix that one later.

**Pull request merged**